Add `#[rustc_confusables]` attribute to allow targeted "no method" error suggestions on standard library types

jieyouxu opened the pull request:
After this PR, the standard library developer can annotate methods on e.g. `BTreeSet::push` with `#[rustc_confusables("insert")]`. When the user mistypes `btreeset.push()`, `BTreeSet::insert` will be suggested if there are no other candidates to suggest. This PR lays the foundations for contributors to add `rustc_confusables` annotations to standard library types for targeted suggestions, as specified in #59450, or to address cases such as #108437.

### Example

Assume `BTreeSet` is the standard library type:

```
// Standard library definition
#![feature(rustc_attrs)]

struct BTreeSet;

impl BTreeSet {
    #[rustc_confusables("push")]
    fn insert(&self) {}
}

// User code
fn main() {
    let x = BTreeSet {};
    x.push();
}
```

A new suggestion (which has lower precedence than suggestions for misspellings and only is shown when there are no misspellings suggestions) will be added to hint the user maybe they intended to write `x.insert()` instead:

```
error[E0599]: no method named `push` found for struct `BTreeSet` in the current scope
  --> test.rs:12:7
   |
3  | struct BTreeSet;
   | --------------- method `push` not found for this struct
...
12 |     x.push();
   |       ^^^^ method not found in `BTreeSet`
   |
help: you might have meant to use `insert`
   |
12 |     x.insert();
   |       ~~~~~~

error: aborting due to previous error
```

Nilstrieb commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -48,6 +51,46 @@ use rustc_hir::intravisit::Visitor;
+}
+
+#[derive(Debug)]
+struct ConfusablesDirectiveParseErr {
```
this should be a session diagnostics like the others in `rustc_hir_typeck/src/errors.rs` (see https://rustc-dev-guide.rust-lang.org/diagnostics/translation.html#writing-translatable-diagnostics)

this way you can easily return it and then just call `sess.emit_err` in the caller.

jieyouxu replied:
You mean something like

```rust
#[derive(Diagnostic)]
#[diag(hir_typeck_invalid_rustc_confusable_attr)]
struct ConfusablesDirectiveParseErr {
    #[primary_span]
    #[label]
    span: Span,
}
```

right?

Nilstrieb replied:
yes

Nilstrieb commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -48,6 +51,46 @@ use rustc_hir::intravisit::Visitor;
+}
+
+#[derive(Debug)]
+enum ConfusablesDirectiveParseErrKind {
+    InvalidArguments,
```
I don't think this needs to be a separate type here unless you add more variants.

Nilstrieb commented on tests/ui/attributes/rustc_confusables.rs:
```diff
@@ -0,0 +1,26 @@
+#![feature(rustc_attrs)]
```
Can you add a second test that checks whether the attribute also works cross-crate? You can add new file in `./auxiliary` and then use `extern crate` here (or rather, in a second test file).
Not all attributes are persisted in metadata across crates so having a test for it is useful (you did it correctly and it should work, if you add `@only_local` to the attribute declaration the test should fail).

jieyouxu replied:
I moved the declaration of and attributes on `BTreeSet` to the auxillary test file (should be counted as another crate right?) and kept the uses to the primary test file.

Noratrieb reviewed (commented):
this may be something that could be made stable in the `#[diagnostic]` namespace in the future (see #111780 and the RFC). cc @weiznich you may be interseted

est31 commented:
The [suggestion](https://github.com/rust-lang/rust/issues/59450#issuecomment-1284399944) made in the thread seems relevant, to use `doc(alias)` instead. Since #107108 it is already used for providing the needed suggestions.

Or is it explicitly wanted that if you search for [`BtreeSet::push`](https://doc.rust-lang.org/stable/std/?search=btreeset%3A%3Apush) then `BtreeSet::insert` should not show up? That seems unfortunate to me.

jieyouxu commented:
> The [suggestion](https://github.com/rust-lang/rust/issues/59450#issuecomment-1284399944) made in the thread seems relevant, to use `doc(alias)` instead. Since #107108 it is already used for providing the needed suggestions.

Oops, I have no clue how I missed that... I will close this PR then.

> Or is it explicitly wanted that if you search for [BtreeSet::push](https://doc.rust-lang.org/stable/std/?search=btreeset%3A%3Apush) then BtreeSet::insert should not show up? That seems unfortunate to me.

I think it's just so happens that `BTreeSet::push` doesn't have a `#[doc(alias)]` for `insert`, so I never noticed while working on this PR.

Sorry about this!

est31 commented:
@jieyouxu note that this is only my private opinion and I don't know if others agree... :) I did not want to discourage you, and it might be too early to close this PR just yet. I think it's definitely worth to discuss having an attribute separate from `#[doc(alias = ...)]`, and it's really great that you provided an implementation, so thanks a lot for that!

I also dimly remember that either the libs (api) team or the rustdoc team expressed that they want to `doc(alias)` to be deployed very sparingly, but I can't find the specific thread, it was not in the tracking issue #50146 at least.

But on the other hand, adding a `push` alias for `insert` seems in-line with #91383 and #97565 which seem to still be present.

jieyouxu commented:
> I think it's definitely worth to discuss having an attribute separate from `#[doc(alias = ...)]`

I suppose I will keep this PR open for now, then, in case having a separate attribute might be useful.

> But on the other hand, adding a push alias for insert seems in-line with https://github.com/rust-lang/rust/pull/91383 and https://github.com/rust-lang/rust/pull/97565 which seem to still be present.

I think I will open a separate PR for adding a doc-alias for `BTreeSet::push` to `BTreeSet::insert`.

est31 commented:
FTR the policy can be found [here](https://std-dev-guide.rust-lang.org/documentation/doc-alias-policy.html) and some links to more discussions can be found [here](https://github.com/rust-lang/libs-team/issues/18).

jieyouxu commented:
Actually, after grepping through `library/std` and `library/core`, I don't think something like adding an alias for `BTreeSet::insert` would pass the `doc(alias)` policy requirement since it would be something like `#[doc(alias = "push")]` but `push` is also used commonly on many other stdlib types like `Vec`? So maybe a separate attribute might make sense after all, which would only affect no-method-found diagnostics and not documentation.

est31 commented:
Fair points... I think ultimately it's up to the libs team if they want `doc(alias)`, `#[diagnostic::confusable]`, or neither.

Nilstrieb commented:
I think the problem with doc aliases is that if you have too many, it becomes quite hard to navigate the docs because you search for something and get lots of random other results showing up. In the diagnostics here you don't really have that problem as its only limited to impls on the type.

WaffleLapkin commented:
r? compiler

compiler-errors commented:
r? compiler

cjgillot commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -48,6 +52,42 @@ use rustc_hir::intravisit::Visitor;
+
+#[derive(Diagnostic)]
+#[diag(hir_typeck_invalid_rustc_confusable_attr)]
+struct ConfusablesDirectiveParseErr {
```
This type should go with the others in `rustc_hir_typeck::errors`.

cjgillot commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -48,6 +52,42 @@ use rustc_hir::intravisit::Visitor;
+}
+
+impl ConfusablesDirective {
+    fn try_parse(attr: &NormalAttr) -> Result<Self, ConfusablesDirectiveParseErr> {
```
```suggestion
    fn try_parse(attr: &NormalAttr) -> Result<ConfusablesDirective, ConfusablesDirectiveParseErr> {
```
Nit: `Self` is better kept for generic contexts.

cjgillot commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -48,6 +52,42 @@ use rustc_hir::intravisit::Visitor;
+}
+
+impl ConfusablesDirective {
+    fn try_parse(attr: &NormalAttr) -> Result<Self, ConfusablesDirectiveParseErr> {
+        let AttrArgs::Delimited(DelimArgs { tokens, .. }) = &attr.item.args else {
+            return Err(ConfusablesDirectiveParseErr {
+                span: attr.item.span(),
+            });
+        };
+
+        let mut names = Vec::new();
+        for tok in tokens.trees() {
+            if let TokenTree::Token(tok, _) = tok
+                && let token::Token { kind, .. } = tok
+                && let token::TokenKind::Literal(lit) = kind
+                && let token::Lit { kind: token::LitKind::Str, symbol, .. } = lit
+            {
+                names.push(symbol.to_string());
+            }
+        }
+
+        Ok(ConfusablesDirective { names })
+    }
```
See `rustc_attribute::builtin` on examples how to parse attributes. For instance `parse_stability`. The token tree should not be directly manipulated.

cjgillot commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1026,6 +1066,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                {
+                                    match ConfusablesDirective::try_parse(attr) {
+                                        Ok(c) => {
+                                            if c.names.contains(&item_name.to_string()) {
```
```suggestion
                                            if c.names.contains(item_name.as_str()) {
```

cjgillot commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1026,6 +1066,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                        Ok(c) => {
+                                            if c.names.contains(&item_name.to_string()) {
+                                                candidate_confusable = Some(inherent_method);
+                                                break;
```
Should this `break` the loop over impl blocks?

cjgillot commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1026,6 +1066,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                            }
+                                        }
+                                        Err(e) => {
+                                            self.sess().emit_err(e);
```
This is not the right place to emit this error. It may be related to an attribute in a different crate.
Also, we don't report malformed attributes in the current crate if they are not used.
Such reporting should go in `rustc_passes::check_attr`.

cjgillot commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1026,6 +1066,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                item_name.span,
+                                format!(
+                                    "you might have meant to use `{}`",
+                                    candidate_confusable.name.as_str()
```
```suggestion
                                    candidate_confusable.name,
```

jieyouxu commented:
@rustbot ready

cjgillot commented:
Could you add a test with malformed `rustc_confusable` attributes, and check that errors are reported irrespective of whether the item is used? Also check that we report an error if the attribute is not on an inherent item?

jieyouxu commented:
> Could you add a test with malformed `rustc_confusable` attributes, and check that errors are reported irrespective of whether the item is used? Also check that we report an error if the attribute is not on an inherent item?

- Added tests to check for malformed `rustc_confusables` attributes (irrespective of whether the item is used):
  - no list at all:`#[rustc_confusable]`
  - empty list: `#[rustc_confusable()]`
  - invalid meta item kind: `#[rustc_confusables(blah)]`
- Added test to check that an error is reported if attribute is not on an inherent item.

WaffleLapkin commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -1031,6 +1031,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            for inherent_method in
+                                self.tcx.associated_items(inherent_impl_did).in_definition_order()
+                            {
+                                if let Some(attr) = self.tcx.get_attr(inherent_method.def_id, sym::rustc_confusables)
+                                    && let Some(candidates) = parse_confusables(attr)
+                                    && candidates.contains(&item_name.name)
```
I think while it makes sense to have `#[rustc_confusables]` because `#[doc(alias)]` pollutes the search, it still makes sense to actually check *both* for diagnostics, so that we don't have to write 2 annotations when `#[doc(alias)]` is already present

jieyouxu replied:
I believe #107108 *already* makes the `#[doc(alias)]` check to make suggestions. `#[rustc_confusables]` shouldn't trigger if any `similar_candidate` was present to make the `#[doc(alias)]` "there is a similar method named ..." suggestion.

jieyouxu commented on compiler/rustc_passes/messages.ftl:
```diff
@@ -276,6 +282,9 @@ passes_expr_not_allowed_in_context =
     the `main` function cannot be declared in an `extern` block
 
+passes_failed_to_extract_meta_item =
+    failed to extract meta item
```
(What are they actually called that's not rustc-internal?)

jieyouxu commented on tests/ui/attributes/rustc_confusables.rs:
```diff
@@ -0,0 +1,43 @@
+    // Misspellings (similarly named methods) take precedence over `rustc_confusables`.
+    let x = BTreeSet {};
+    x.inser();
+    //~^ ERROR no method named
```
For the "there is a method with a similar name: `specific_name`", unfortunately due to how `compiletest` handles the actual errors, I can't seem to match on `specific_name`, best I can do is

```
//~| HELP there is a method with a similar name
```

jieyouxu replied:
For the ``"there is a method with a similar name: `specific_name`"``, unfortunately due to how `compiletest` handles the actual errors, I can't seem to match on `specific_name`, best I can do is

```
//~| HELP there is a method with a similar name
```

bors commented:
:umbrella: The latest upstream changes (presumably #113328) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -1986,6 +1987,42 @@ impl CheckAttrVisitor<'_> {
+    fn check_confusables(&self, attr: &Attribute, target: Target) -> bool {
+        match target {
+            Target::Method(MethodKind::Inherent) => {
+                let Some(meta) = attr.meta() else {
+                    self.tcx.sess.emit_err(errors::FailedToExtractMetaItem { span: attr.span });
+                    return false;
+                };
+                let ast::MetaItem { kind: MetaItemKind::List(ref metas), .. } = meta else {
+                    self.tcx.sess.emit_err(errors::InvalidMetaItemKind { span: attr.span });
+                    return false;
+                };
```
I don't think you need to emit an error in those cases, just early return. You already get a parser error.

cjgillot commented on compiler/rustc_feature/src/builtin_attrs.rs:
```diff
@@ -625,6 +625,12 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
     ),
+    rustc_attr!(
+        rustc_confusables, Normal,
+        template!(List: "name1, name2, ..."),
```
Maybe quote the names, as that's what we expect?

cjgillot commented on tests/ui/attributes/rustc_confusables.stderr:
```diff
@@ -0,0 +1,69 @@
+LL |     #[rustc_confusables]
+   |     ^^^^^^^^^^^^^^^^^^^^
+
+error[E0539]: incorrect meta item
+  --> $DIR/rustc_confusables.rs:40:25
+   |
+LL |     #[rustc_confusables(invalid_meta_item)]
+   |                         ^^^^^^^^^^^^^^^^^
+
```
Should we say that we expect a quoted string, and suggest quotes?

jieyouxu replied:
Added a suggestion to surround it in quotes.

cjgillot commented on tests/ui/attributes/rustc_confusables.stderr:
```diff
@@ -0,0 +1,69 @@
+LL |     #[rustc_confusables()]
+   |     ^^^^^^^^^^^^^^^^^^^^^^
+
+error: invalid meta item kind
+  --> $DIR/rustc_confusables.rs:34:5
+   |
+LL |     #[rustc_confusables]
+   |     ^^^^^^^^^^^^^^^^^^^^
+
```
Not useful, as we already have an error at line 1 of stderr.

jieyouxu committed:
Add infrastructure `#[rustc_confusables]` attribute to allow targeted
"no method" errors on standard library types

The standard library developer can annotate methods on e.g.
`BTreeSet::push` with `#[rustc_confusables("insert")]`. When the user
mistypes `btreeset.push()`, `BTreeSet::insert` will be suggested if
there are no other candidates to suggest.

cjgillot commented on compiler/rustc_hir_typeck/messages.ftl:
```diff
@@ -52,6 +52,7 @@ hir_typeck_functional_record_update_on_non_struct =
 
 hir_typeck_help_set_edition_cargo = set `edition = "{$edition}"` in `Cargo.toml`
 hir_typeck_help_set_edition_standalone = pass `--edition {$edition}` to `rustc`
+
```
stray change.

cjgillot reviewed (commented):
Thanks! 3 nits and r=me.

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 08c77a6eb403768ed64ae567980887984f506ef9 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 08c77a6eb403768ed64ae567980887984f506ef9 -->
<!-- homu: {"type":"Approved","sha":"08c77a6eb403768ed64ae567980887984f506ef9","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 08c77a6eb403768ed64ae567980887984f506ef9 with merge c4b1a20fd4e7f3fa31ac9fbf4b3e8dad1f8a912b...
<!-- homu: {"type":"BuildStarted","head_sha":"08c77a6eb403768ed64ae567980887984f506ef9","merge_sha":"c4b1a20fd4e7f3fa31ac9fbf4b3e8dad1f8a912b"} -->

rust-log-analyzer commented:
The job **`armhf-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5567750147/jobs/10169869152) [(plain)](https://github.com/rust-lang-ci/rust/commit/c4b1a20fd4e7f3fa31ac9fbf4b3e8dad1f8a912b/checks/15077590228/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5567750147/jobs/10170131635)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/5567750147/jobs/10170131635","builder_name":"checks-actions"} -->

cjgillot commented:
Seems spurious: `client.read_exact(&mut header) failed with Connection reset by peer`
@bors retry

bors commented:
:hourglass: Testing commit 08c77a6eb403768ed64ae567980887984f506ef9 with merge 11da267fdb17371d60e6f555d4bcf35bf765d982...
<!-- homu: {"type":"BuildStarted","head_sha":"08c77a6eb403768ed64ae567980887984f506ef9","merge_sha":"11da267fdb17371d60e6f555d4bcf35bf765d982"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5568430510/jobs/10171780911)
Approved by: cjgillot
Pushing 11da267fdb17371d60e6f555d4bcf35bf765d982 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5568430510/jobs/10171780911"},"merge_sha":"11da267fdb17371d60e6f555d4bcf35bf765d982"} -->

rust-timer commented:
Finished benchmarking commit (11da267fdb17371d60e6f555d4bcf35bf765d982): [comparison URL](https://perf.rust-lang.org/compare.html?start=4a07b2baf500e6619b4106e90a242e2ecacf3dd0&end=11da267fdb17371d60e6f555d4bcf35bf765d982&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=4a07b2baf500e6619b4106e90a242e2ecacf3dd0&end=11da267fdb17371d60e6f555d4bcf35bf765d982&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4a07b2baf500e6619b4106e90a242e2ecacf3dd0&end=11da267fdb17371d60e6f555d4bcf35bf765d982&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6% | [0.6%, 0.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.6% | [0.6%, 0.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4a07b2baf500e6619b4106e90a242e2ecacf3dd0&end=11da267fdb17371d60e6f555d4bcf35bf765d982&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=4a07b2baf500e6619b4106e90a242e2ecacf3dd0&end=11da267fdb17371d60e6f555d4bcf35bf765d982&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 658.073s -> 657.937s (-0.02%)
